Coral/Coral.js

```diff
@@ -0,0 +1,312 @@
+  const { x: u, y: v } = point;
+
+  return {
+    x: QUAD_X + u * QUAD_WIDTH,
+    y: QUAD_Y + (1 - v) * QUAD_HEIGHT,
+  };
+}
+
+function clamp(x, min, max) {
+  return Math.max(Math.min(x, max), min);
+}
+
+class Rect {
+  constructor(x, y, width, height) {
+    this.position = { x, y };
+    this.dimensions = { x: width, y: height };
+  }
+
+  clamp(point) {
+    const { x, y } = point;
+
+    return {
+      x: clamp(x, this.position.x, this.position.x + this.dimensions.x),
+      y: clamp(y, this.position.y, this.position.y + this.dimensions.y),
+    };
+  }
+}
+
+class ControlPoint {
+  constructor(x, y, dx, dy) {
+    this._position = { x, y };
+    this._tangent = { x: dx, y: dy };
+
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+
+  get position() {
+    return this._position;
+  }
+
+  set position(value) {
+    this._position = value;
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+
+  get tangent() {
+    return this._tangent;
+  }
+
+  set tangent(value) {
+    this._tangent = value;
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+}
+
+class Spline {
+  constructor(control_points) {
+    this.control_points = control_points;
+  }
+
+  log_params() {
+    console.log("current control points:");
+    for (const point of this.control_points) {
+      console.log(
+        "p:",
+        point.position.x.toPrecision(3),
+        point.position.y.toPrecision(3)
+      );
+      console.log(
+        "v:",
+        point.tangent.x.toPrecision(3),
+        point.tangent.y.toPrecision(3)
+      );
+    }
+  }
+}
+
+const SELECT_RADIUS = 8;
+const SELECT_RADIUS_SQR = SELECT_RADIUS * SELECT_RADIUS;
+
+class InteractiveVertex {
+  constructor(control_point, constraint) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+  }
+
+  get position() {
+    return this.control_point.position;
+  }
+
+  is_hovering(mouse) {
+    const position_world = uv_to_world(this.control_point.position);
+    const dist_sqr = norm(sub(mouse, position_world));
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(uv) {
+    this.control_point.position = this.constraint.clamp(uv);
+  }
+}
+
+class InteractiveTangent {
+  constructor(control_point, constraint) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+  }
+
+  get tip() {
+    return this.control_point.forward_point;
+  }
+
+  is_hovering(mouse) {
+    const position_world = uv_to_world(this.control_point.forward_point);
+    const dist_sqr = norm(sub(mouse, position_world));
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(uv) {
+    const tangent = sub(uv, this.control_point.position);
+    this.control_point.tangent = this.constraint.clamp(tangent);
+  }
+}
+
+const SPLINE = new Spline([
```

Comment:
For the next step I'll need to make a `CoralTile` class that stores 4 control points. And also I need a way to encode the 16 tile types.